### PR TITLE
CORE-4117: Clean up to allow planners perform action which switched t…

### PR DESCRIPTION
**Pull request by skaynama**
…o (helps with detecting loop of death and invoking recovery for docking planner).

### CODE REVIEW

**Comment by skaynama** [comment]
@jasonimercer, @p6chen

**Comment by skaynama** [comment]
Accompanying PRs in cpr_pathtracker and cpr_base_navigation.

File: move_base/src/move_base.cpp [hunk 1]
```diff
@@ -1096,16 +1103,17 @@ namespace move_base {
...
           vel_pub_.publish(cmd_vel);
 
-          // It is possible for computeVelocityCommands to return true when we are waiting for dynamics
+          // It is possible for computeVelocityCommands to return true when we are waiting for dynamic costmap logic
           // to timeout. In that case, custom_status == nav_core::status::WAIT. If we are in an OK state
           // where meaningful cmd_vel is being published then we can reset the indices here in move_base
           // as well as in the recovery_manager.
           if (custom_status == nav_core::status::OK)
           {
+            last_valid_control_ = ros::Time::now();
```
Comment (skaynama) [inline move_base/src/move_base.cpp]: Setting this variable here instead prevents it from being set while ```custom_state``` is ```WAIT```. That way if no valid control can be found within the ```controller_patience``` we can still go to recovery (currently that block is never executed due to this variable being set unintentionally).

Accompanying PR increase ```controller_patience``` to beyond dynamic costmap wait time.

File: nav_core/include/nav_core/base_global_planner.h [hunk 2]
```diff
@@ -73,7 +73,12 @@ namespace nav_core {
       virtual void initialize(std::string name, costmap_2d::Costmap2DROS* costmap_ros) = 0;
 
       /**
-       * @brief Function to reset the planner state, whatever that mean for each plugin.
+       * @brief Function to prepare the planner for actions post-recovery, whatever that means for each plugin.
+       */
+      virtual void prepareForPostRecovery(){}
```
Comment (skaynama) [inline nav_core/include/nav_core/base_global_planner.h]: New function to separate out actions that are to be performed in preperation for recovery (old behaviour that was done in ```resetPlanner()```) and what actually resets the planner.

```resetPlanner()``` is now used to perform actions (as dictated by each planner) every time the global planner is switched to.

**Comment by jasonimercer** [comment]
LGTM. I like that the method name now describes what is being done.

**Pull request merged**